Deploy Apache Spark on Google Axion C4A VM (#9)
odidev/arm-learning-paths: content/learning-paths/servers-and-cloud-computing/spark-on-gcp/benchmarking.md

Reviewer: ## **Spark Shell** is an interactive command-line tool provided by Apache Spark that lets you run Spark jobs interactively using Scala (or Python via pyspark).
## While you can use spark-shell for basic timing (like with System.nanoTime() you did in your benchmarking), it's not meant for serious benchmarking.

## I recommend using tools like **spark-sql-perf, HiBench** for the proper benchmarking. Use GPT for good benchmarking examples.

---
title: Benchmarking
weight: 5

### FIXED, DO NOT MODIFY
layout: learningpathall
---

## Apache Spark Internal Benchmarking
Apache Spark includes internal micro-benchmarks to evaluate the performance of core components like SQL execution, aggregation, joins, and data source reads. These benchmarks are helpful for comparing platforms such as x86_64 vs Arm64.
Below are the steps to run Spark’s built-in SQL benchmarks using the SBT-based framework.

1. Clone the Apache Spark source code
```console
$ git clone https://github.com/apache/spark.git
```
This downloads the full Spark source including internal test suites and the benchmarking tools.

2. Checkout the desired Spark version
```console
$ cd spark/ && git checkout v4.0.0
```

> Switche to the stable Spark 4.0.0 release, which supports the latest internal benchmarking APIs.

Reviewer: ```suggestion
Switch to the stable Spark 4.0.0 release, which supports the latest internal benchmarking APIs.
```

[info] Java HashMap (two ints)                             273            277           2         76.7          13.0       0.5            X
[info] Java HashMap (UnsafeRow)                            507            510           3         41.4          24.2       0.3            X
[info] LongToUnsafeRowMap (opt=false)                      237            238           0         88.3          11.3       0.6            X
[info] LongToUnsafeRowMap (opt=true)                        76             77           1        277.1           3.6       1.8            X
[info] BytesToBytesMap (off Heap)                          472            474           2         44.4          22.5       0.3            X
[info] BytesToBytesMap (on Heap)                           505            506           1         41.6          24.1       0.3            X
[info] Aggregate HashMap                                    23             23           0        913.0           1.1       5.9            X
[success] Total time: 669 s (11:09), completed Jul 24, 2025, 5:41:24 AM

```
### Benchmark summary on x86_64:

> The following  benchmark results are collected on an x86_64 environment.

Reviewer: ```suggestion
The following  benchmark results are collected on a c3-standard-4 (4 vCPU, 2 core, 16 GB Memory) x86_64 environment.
```

| BytesToBytesMap           | Java HashMap (UnsafeRow)             | 544                | 546               | 2              | 38.5           | 26.0             | 0.4X         |
| BytesToBytesMap           | LongToUnsafeRowMap (opt=false)       | 352                | 355               | 1              | 59.5           | 16.8             | 0.6X         |
| BytesToBytesMap           | LongToUnsafeRowMap (opt=true)        | 74                 | 75                | 1              | 284.6          | 3.5              | 2.8X         |
| BytesToBytesMap           | BytesToBytesMap (off Heap)           | 623                | 628               | 7              | 33.7           | 29.7             | 0.3X         |
| BytesToBytesMap           | BytesToBytesMap (on Heap)            | 624                | 627               | 3              | 33.6           | 29.8             | 0.3X         |
| BytesToBytesMap           | Aggregate HashMap                    | 31                 | 31                | 0              | 680.7          | 1.5              | 6.6X         |

### Benchmark summary on Arm64:

> The following  benchmark results are collected on an Arm64 environment using the C4A virtula machine series (c4a-standard-4) on GCP.

Reviewer: ```suggestion
The following  benchmark results are collected on a c4a-standard-4 (4 vCPU, 16 GB Memory) Arm64 environment, running RHEL 9.
```

| Max function bytecode size | codegen = F              | 263            | 268            | 4          | 2.5         | 401.6          | 1.0       |
| Max function bytecode size | hugeMethodLimit = 10000  | 143            | 148            | 8          | 4.6         | 217.4          | 1.8       |
| Max function bytecode size | hugeMethodLimit = 1500   | 129            | 132            | 3          | 5.1         | 196.6          | 2.0       |
| Cube                       | wholestage off           | 1572           | 1582           | 14         | 3.3         | 299.9          | 1.0       |
| Cube                       | wholestage on            | 841            | 843            | 2          | 6.2         | 160.4          | 1.9       |
| BytesToBytesMap            | UnsafeRowhash            | 137            | 137            | 0          | 153.6       | 6.5            | 1.0       |
| BytesToBytesMap            | murmur3 hash             | 48             | 48             | 0          | 440.6       | 2.3            | 2.9       |
| BytesToBytesMap            | fast hash                | 42             | 42             | 0          | 499.2       | 2.0            | 3.3       |

> | Aggregate HashMap          |                          | 23             | 23             | 0          | 913.0       | 1.1            | 5.9       |

Reviewer: verify this blank cell.

---
title: Benchmarking
weight: 5

### FIXED, DO NOT MODIFY
layout: learningpathall
---

## Apache Spark Internal Benchmarking (v4.0.0)
Apache Spark includes internal micro-benchmarks to evaluate the performance of core components like SQL execution, aggregation, joins, and data source reads. These benchmarks are helpful for comparing platforms such as x86_64 vs Arm64.
Below are the steps to run Spark’s built-in SQL benchmarks using the SBT-based framework.

> Clone the Apache Spark source code

Reviewer: ```suggestion
1. Clone the Apache Spark source code
```

```console
$ git clone https://github.com/apache/spark.git
```

> Downloads the full Spark source including internal test suites and benchmarking tools.

Reviewer: ```suggestion
This downloads the full Spark source including internal test suites and the benchmarking tools.
```

Checkout the desired Spark version
```console
$ cd spark/ && git checkout v4.0.0
```
Switches to the stable Spark 4.0.0 release, which supports the latest internal benchmarking APIs.

> Build Spark with benchmarking profile enabled

Reviewer: ```suggestion
3. Build Spark with benchmarking profile enabled
```

```console
$ ./build/sbt -Pbenchmarks clean package
```
Compiles Spark and its dependencies, enabling the benchmarks build profile for performance testing.

Run a built-in benchmark suite
```console
$ ./build/sbt -Pbenchmarks "sql/test:runMain org.apache.spark.sql.execution.benchmark.AggregateBenchmark"
```  

> Executes the AggregateBenchmark, which compares performance of SQL aggregation operations (e.g., SUM, STDDEV) with and without WholeStageCodegen.

Reviewer: ```suggestion
This executes the AggregateBenchmark, which compares performance of SQL aggregation operations (e.g., SUM, STDDEV) with and without WholeStageCodegen. WholeStageCodegen is an optimization technique used by Spark SQL to improve the performance of query execution by generating Java bytecode for entire query stages (aka whole stages) instead of interpreting them step-by-step.
```
